### PR TITLE
Add basic classification fine-tuning capabilities for Geneformer

### PR DESCRIPTION
**Pull request by mattwoodx**
- Addition of the `fine_tune_classifier` method to the `HelicalBaseFoundationModel`.
- Loads the current model that the user has selected in the config for fine-tuning.
- Returns a BertForSequenceClassification model.
- Addition of some basic testing for the method.

### CODE REVIEW

File: helical/models/geneformer/geneformer_utils.py [hunk 1]
```diff
@@ -220,4 +221,112 @@ def mean_nonpadding_embs(embs, original_lens, dim=1):
     elif embs.dim() == 2:
         masked_embs = embs.masked_fill(~mask, 0.0)
         mean_embs = masked_embs.sum(dim) / original_lens.float()
-    return mean_embs
\ No newline at end of file
+    return mean_embs
+
+# fine tune Geneformer for classification tasks
+def classification_fine_tuning(
```
Comment (JadSbai) [inline helical/models/geneformer/geneformer_utils.py:227]: Super clean and clear. Love it.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -35,7 +35,7 @@ jobs:
           pytest --cov-report=html:html_cov --cov-branch --cov-report term --cov=helical ci/
       
       - name: Upload coverage report
-        uses: actions/upload-artifact@v2
+        uses: actions/upload-artifact@v4
```
Comment (JadSbai) [inline .github/workflows/main.yml:38]: Yeah got the same issue on my branch. Need to make sure this is updated on main afterwards

File: ci/tests/test_geneformer/test_geneformer_model.py [hunk 3]
```diff
@@ -45,4 +47,11 @@ def test_ensembl_data_is_caught(self, gene_symbols, raises_error):
         if raises_error:
             with pytest.raises(ValueError):
                 self.geneformer.process_data(self.data, "gene_symbols")
+
+    def test_fine_tune_classifier(self):
+        assert self.geneformer.fine_tune_classifier is not None
+        fine_tuned_model = self.geneformer.fine_tune_classifier(self.tokenized_dataset)
+        assert fine_tuned_model is not None
+        assert fine_tuned_model is not self.geneformer.model
+        assert isinstance(fine_tuned_model, BertForSequenceClassification)
```
Comment (JadSbai) [inline ci/tests/test_geneformer/test_geneformer_model.py:56]: Are you sure this is checking that the weights of the fine-tuned model have been updated indeed? got a feeling that is might only be checking the class (on line 55 i mean). Might be wrong though.

File: helical/models/geneformer/model.py [hunk 4]
```diff
@@ -186,3 +187,59 @@ def ensure_data_validity(self, adata: AnnData, gene_names: str) -> None:
...
+            train_dataset: Dataset, 
+            optimizer: optim = optim.AdamW,
+            optimizer_params: dict = {'lr': 0.0001}, 
+            loss_function: loss = loss.CrossEntropyLoss(), 
+            label: str = "cell_types", 
+            epochs: int = 1,
+            freeze_layers: int = 0,
+            validation_dataset: Optional[Dataset] = None,
+            lr_scheduler_params: Optional[dict] = None) -> BertForSequenceClassification:
+        """Fine-tunes the Geneformer model for classification tasks.
```
Comment (JadSbai) [inline helical/models/geneformer/model.py:202]: Ok, this is me nitpicking, but I think it'd be nice if you could add the Return value section as part of the doc strings as well as properly specifying the default value for optimizer, loss function etc. Cause I think the docs are then automatically generated based on these (so need to be formatted properly). Also, it'd be cool to explain why we're returning a BertForSequenceClassification model. End of my nitpicking.

**Review by JadSbai — COMMENTED** [review]
Super clean. Nothing much to add, just a few minor comments here and there.

File: ci/tests/test_geneformer/test_geneformer_model.py (hunk 3 above)
Comment (mattwoodx) [reply ci/tests/test_geneformer/test_geneformer_model.py:56]: It doesn't do a weight check. At the moment it is just checking that it is not the original model, that we have instantiated a new model for the fine tuning and that the return type is correct.

File: helical/models/geneformer/model.py (hunk 4 above)
Comment (mattwoodx) [reply helical/models/geneformer/model.py:202]: Updated!

**Comment by maxiallard** [comment]
Thanks for adding the function to the geneformer model. Since we are still iterating on the fine-tuning function (and the way we want to include it  for other models, I would wait to have all the fine-tuning functions defined for each RNASeq model for now before merging it into main.

**Comment by bputzeys** [comment]
I agree with Maxime. Maybe now is a good point to adjust the base class all the models inherit from to have a function called 'fine_tune_model'. The other models can, for now, just have a pass and 'NotImplemented' Error as implementation. But then the structure is there.

**Comment by bputzeys** [comment]
Out of curiosity where did you get the code from or did you do it all by yourself?
